docs/docs/segment-strava.md

```diff
@@ -0,0 +1,123 @@
+---
+id: strava
+title: Strava
+sidebar_label: Strava
+---
+
+## What
+
+[Strava][strava] ia a popular activity tracker for bike, run or any other traning.
+ To keep up with your training goals it is important to be constantly reminded about it.
+
+A Strava Oh My Posh segment would show your last activity,
+ and also indicate by a color if it is time to get away from your computer and do some workout.
+
+## Accessing your Strava data
+
+There are several ways you could get your last activity.
+
+### Build you own api, integrating with Strava
```

Comment:
This won't work for most people so we can't do it like this. We have to access the API directly to get the latest activity rather than forcing people to become full blown programmers (which a lot won't be able to do)

The following API endpoints lists all activities for a given athlete:
`https://www.strava.com/api/v3/athletes/id/activities`

The only challenge is then getting the actual token as it needs to be refreshed in a timely fashion. That's maybe something we can facilite by having that logic inside oh-my-posh by means of a link you can click so you can provide an `access_token` and `refresh_token` for us to store.

Reply:
Yes, that would absolute be a more user friendly approach.  I think that to access strava api we have two options:

1. Ask users to create an app in the strava developer portal
2. Someone else (you or me) creates an app in strava developer portal

This app must be linked to a url to allow for login (and then get the access and refresh tokens). The user will have to grant access to that url also.

Do you think adding a https://ohmyposh.dev/auth/strava is a good idea? That would make it much easier for future segments that need this type of data.  I'm not sure if I'm up for that job.

As an alternative I woke up my old project `fartslek.no` and provided a simple endpoint that returns the last activity  by a token. I did have plans for this, so just now, you may go and try out fartslekreg for a simple way of getting your last activity.

I've also updated the docs.

Reply:
Yes, I would create that path. We have the ability to add an [API endpoint](https://docs.microsoft.com/en-us/azure/static-web-apps/add-api?tabs=vanilla-javascript) to the docs as it's an Azure Static web site. This has got quite the information, maybe I'll give that a go but the site is part of this codebase under /docs so anyone could add it.

Reply:
@larserikfinholt I just pushed a dummy one at https://github.com/JanDeDobbeleer/oh-my-posh/pull/1536

Reply:
@JanDeDobbeleer  Just to make sure I understand:

* The user opens https://ohmyposh.dev/auth/login
* This page shows a "login" button
* User clicks this button, is redirected to strava login page
* Strava then ask for grant to access the `activity:read` scope (and is asked "do you grant access to OhMyPoshStravaSegment https://ohmyposh.dev")
* User accepts, and is redirected to https://ohmyposh.dev/auth/strava
* This page then display the access token, and refresh token, and ask user to copy to clipboard
* User should then maybe store that in a temp file (at same location as the config file?) with write permissions
* Then the segment uses the access token until it expires, and saves the new refresh token and expiry date back to the temp file

Correct?

Reply:
I would indeed construct the authentication first in a separate PR. The way I see it we can redirect directly to Strava, no need for a button as we'll have to trigger the user to click the link from the prompt anyways (we can print it or make a clickable link, let's figure that out afterwards). Ideally the landing page had a copy-paste snippet at the ready, but also something we can make better once it works. Let's assume you're the first and probably only user so we can take our time 😬

I already have a Strava API client created, if we build the function (we'll need 2, one for the initial auth and another one to fetch the redirect) I can add the credentials in Azure.

Reply:
Ok, good. Could you let me know when you have decided on the format on the clipboard? I suggest something like this:

```
'{"access_token":"ACCESS_TOKEN","refresh_token":"REFRESH_TOKEN","expires_at":123123132,"atheleteId":123123}'
```

Reply:
@larserikfinholt I would only create two properties, `access_token` and `refresh_token`. We can try to refresh, it it fails we'll display the link to authenticate again.

Reply:
@JanDeDobbeleer agree. I'll assume the user will receive this from you, and save it in the config file `'{"access_token":"ACCESS_TOKEN","refresh_token":"REFRESH_TOKEN"}'`

Any guidance on the where I should store the updated tokens? Something like this:
```
tokensPath = fmt.Sprintf("%s/.strava/tokens", a.env.homeDir())
```

Reply:
We have a cache you can use in Environment. If you set the timeout to the expiration of the refresh token, we immediately have a way to detect if we no longer have credentials.

Reply:
@larserikfinholt I finally debugged the whole thing locally :-) I have a working api setup that prints the access and refresh token using the oh-my-posh client. Debugging a Static Web App API endpoint in Azure is in beta it seems and the Azure version still gives a 500 (and have no idea why just yet).

Reply:
@JanDeDobbeleer  great! 

I've just started implementing the refresh logic, and relized that I'll be needing the ClientId and ClientSecret to do the refresh.  Do you have some built in logic where I could get those? Some global setting?

Reply:
Allow me to also implement an endpoint for that.

Reply:
I was going MENTAL: https://github.com/Azure/static-web-apps/issues/165

Reply:
@larserikfinholt [link](https://www.strava.com/oauth/authorize?client_id=76033&response_type=code&redirect_uri=https://ohmyposh.dev/api/auth&approval_prompt=force&scope=read,activity:read
)

Reply:
@JanDeDobbeleer  ok. I just completed a (untested) implementation of the refresh logic, you may use it.

Reply:
I would use two properties, just like they are there.

```json
{
  "type": "strava",
  "properties": {
    "access_token": "11111111111111111",
    "refresh_token": "11111111111111111"
  }
}
```

Reply:
@JanDeDobbeleer just pushed a implementation, but it does not work - I get a "invalid token" error from Strava when using the access token I receive from your api. Any idea what could be wrong?

(as you can see, I've removed the "getNewAccessToken" and is always using the one from config - but that should be valid for 6 hours, so I don't think thats the problem)

About the token exchange - should we just reuse the existing api from `https://www.strava.com/api/v3/oauth/token` ? and place it at `https://ohmyposh.dev/api/strava/token` (of course without the need to post clientId and clientSecret)

Reply:
@larserikfinholt  I just tested the token received from @JanDeDobbeleer  link in postman and it works(I get my stats). 
![image](https://user-images.githubusercontent.com/1829553/148191855-9c609f22-3727-4b66-a588-c1b962aa094a.png)

Reply:
@lnu Cool! Glad to see its working. Yes, I fixed a small typo in auth header :-)

Reply:
Updated with tests - only thing missing is to enable the use tokenservice when @JanDeDobbeleer  is ready.

Reply:
I'm back so I'll have a look ;-)

Reply:
refresh is live at `https://ohmyposh.dev/api/refresh?segment=strava&token=refresh_token`

Reply:
@larserikfinholt if you can mark this PR as "Allow edits from maintainers" (can't see if that's enabled or not). See [here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for more info.

Reply:
@JanDeDobbeleer yes, I it's already enabled. Please feel free do do anything :-) (Hopefully I'll get the chance do some updates tonight after work - if you've not already fixed it)

Reply:
I'll add commits on top so you can see the changes 😉